be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/GetMessagesByIdResponseHandler.scala

```diff
@@ -1,86 +1,170 @@
+      implicit builder: GraphDSL.Builder[NotUsed] =>
+        {
+          import GraphDSL.Implicits._
 
-        /* glue the components together */
-        handlerPartitioner.out(portPipelineError) ~> handlerMerger
-        handlerPartitioner.out(portResponseHandler) ~> responseHandler ~> handlerMerger
+          /* partitioner port numbers */
+          val portPipelineError = 0
+          val portResponseHandler = 1
+          val totalPorts = 2
 
-        /* close the shape */
-        FlowShape(handlerPartitioner.in, handlerMerger.out)
-      }
-  })
+          /* building blocks */
+          val handlerPartitioner = builder.add(Partition[GraphMessage](
+            totalPorts,
+            {
+              case Right(jsonRpcMessage: JsonRpcResponse) => jsonRpcMessage.result match {
+                  case Some(result) => result.resultMap match {
+                      case Some(_) => portResponseHandler
+                      case _       => portPipelineError
+                    }
+                  case _ => portPipelineError
+                }
+              case _ => portPipelineError // Pipeline error goes directly in handlerMerger
+            }
+          ))
+
+          val responseHandler = builder.add(GetMessagesByIdResponseHandler.responseHandler(mediatorActorRef, messageRegistry)(system))
+          val handlerMerger = builder.add(Merge[GraphMessage](totalPorts))
+
+          /* glue the components together */
+          handlerPartitioner.out(portPipelineError) ~> handlerMerger
+          handlerPartitioner.out(portResponseHandler) ~> responseHandler ~> handlerMerger
 
-  private def responseHandler(dbActorRef: ActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
+          /* close the shape */
+          FlowShape(handlerPartitioner.in, handlerMerger.out)
+        }
+    })
+
+  // This function packs each message into a publish before pushing into the pipeline
+  private def responseHandler(mediatorActorRef: ActorRef, messageRegistry: MessageRegistry)(implicit system: ActorSystem): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map {
     case Right(jsonRpcMessage: JsonRpcResponse) =>
-      val receivedResponse = jsonRpcMessage.result.get.resultMap.get
-      receivedResponse.keys.foreach(channel => {
-        receivedResponse(channel).foreach(message => {
-          writeOnDb(channel, message, dbActorRef, MAX_ATTEMPT)
-        })
-      })
-      Right(jsonRpcMessage)
+      val validatorFlow = validator(mediatorActorRef, messageRegistry)
+      val receivedResponse: Map[Channel, Set[GraphMessage]] =
+        jsonRpcMessage.result.get
+          .resultMap.get
+          .map {
+            case (channel, set) => (channel, set.map(msg => Right(JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.PUBLISH, new Publish(channel, msg), Some(0)))))
+          }
+
+      val failedMessages = passThroughPipeline(receivedResponse, validatorFlow, MAX_RETRY)
```

Comment:
The map of failed messaged isn't used later on except for checking if it is empty to decide if it succeeded or not. Therefore it could make more sense to either return a simple boolean instead.

Reply:
Absoluetely, fixed !